File: site/js/pdfread.js
// Minimal read-only PDF access. Mirrors clo2plt/pdfread.py.
//
// The file is held twice: as bytes, and as a latin-1 string where one char is
// exactly one byte, so string offsets and byte offsets are interchangeable and
// the Python regexes port across unchanged.

export class PdfError extends Error {}

const OBJ_RE = /(\d+)\s+(\d+)\s+obj\b/g;

export class Pdf {
  constructor(bytes) {
    this.bytes = bytes;
    this.text = new TextDecoder("latin1").decode(bytes);
    if (!this.text.startsWith("%PDF-")) {
      throw new PdfError("Not a PDF file (missing %PDF- header).");
    }
    if (this.text.includes("/ObjStm")) {
      throw new PdfError(
        "This PDF uses object streams, which this reader does not support. " +
          "Re-export it from CLO 3D, or flatten it first with Ghostscript."
      );
    }
    this.offsets = new Map();
    OBJ_RE.lastIndex = 0;
    let m;
    while ((m = OBJ_RE.exec(this.text)) !== null) {
      // Later definitions win: incremental updates append revised objects.
      this.offsets.set(Number(m[1]), m.index + m[0].length);
    }
    this.streams = new Map();
  }

  raw(num) {
    const start = this.offsets.get(num);
    if (start === undefined) throw new PdfError(`object ${num} not found`);
    const end = this.text.indexOf("endobj", start);
    if (end < 0) throw new PdfError(`object ${num} has no endobj`);
    return this.text.slice(start, end);
  }

  // Decode every stream up front so the interpreter can stay synchronous;
  // DecompressionStream is async and the interpreter recurses through forms.
  async decodeAll() {
    this.failed = [];
    for (const num of this.offsets.keys()) {
      const body = this.raw(num);
      const i = body.indexOf("stream");
      if (i < 0) continue;
      try {
        this.streams.set(num, await this.#decode(num, body, i));
      } catch (err) {
        this.failed.push(num);
      }
    }
  }

  async #decode(num, body, i) {
    const base = this.offsets.get(num);
    let j = i + "stream".length;
    if (body.slice(j, j + 2) === "\r\n") j += 2;
    else if (body[j] === "\r" || body[j] === "\n") j += 1;
    let k = body.lastIndexOf("endstream");
    if (k < 0) k = body.length;

    const header = body.slice(0, i);

    // Prefer the declared /Length. DecompressionStream rejects trailing bytes,
    // and there is almost always an EOL between the data and "endstream" --
    // Python's decompressobj ignores it, the browser's does not.
    const lenMatch = /\/Length\s+(\d+)(?!\s+\d+\s+R)/.exec(header);
    let end = base + k;
    if (lenMatch) {
      const declared = base + j + Number(lenMatch[1]);
      if (declared <= base + k) end = declared;
    }

    let payload = this.bytes.subarray(base + j, end);
    if (!header.includes("/FlateDecode")) {
      if (header.includes("/Filter")) {
        throw new PdfError(`object ${num}: unsupported stream filter`);
      }
      return new TextDecoder("latin1").decode(payload);
    }

    try {
      return await inflate(payload);
    } catch (err) {
      // Fall back to trimming trailing EOL bytes, for writers whose /Length
      // disagrees with the actual payload.
      let stop = base + k;
      while (stop > base + j) {
        const c = this.bytes[stop - 1];
        if (c === 0x0a || c === 0x0d || c === 0x20) stop -= 1;
        else break;
      }
      return await inflate(this.bytes.subarray(base + j, stop));
    }
  }

  stream(num) {
    const s = this.streams.get(num);
    if (s === undefined) throw new PdfError(`object ${num} has no usable stream`);
    return s;
  }

  hasStream(num) {
    return this.streams.has(num);
  }

  ref(body, key) {
    const m = new RegExp(`/${key}\\s+(\\d+)\\s+\\d+\\s+R\\b`).exec(body);
    return m ? Number(m[1]) : null;
  }

  name(body, key) {
    const m = new RegExp(`/${key}\\s*\\(`).exec(body);
    if (!m) return null;
    return unescapeString(balancedString(body, m.index + m[0].length - 1));
  }

  numbers(body, key) {
    const m = new RegExp(`/${key}\\s*\\[([^\\]]*)\\]`).exec(body);
    if (!m) return null;
    const found = m[1].match(/[-+]?(?:\d+\.?\d*|\.\d+)(?:[eE][-+]?\d+)?/g);
    return found ? found.map(Number) : [];
  }

  xobjects(body) {
    const out = {};
    const re = /\/([A-Za-z0-9_.#-]+)\s+(\d+)\s+\d+\s+R/g;
    let m;
    while ((m = re.exec(body)) !== null) out[m[1]] = Number(m[2]);
    return out;
  }
}

async function inflate(payload) {
  const stream = new Blob([payload])
    .stream()
    .pipeThrough(new DecompressionStream("deflate"));
  const buf = await new Response(stream).arrayBuffer();
  return new TextDecoder("latin1").decode(new Uint8Array(buf));
}

function balancedString(body, start) {
  let depth = 0;
  let i = start;
  let out = "";
  while (i < body.length) {
    const c = body[i];
    if (c === "\\") {
      out += body.slice(i, i + 2);
      i += 2;
      continue;
    }
    if (c === "(") {
      depth += 1;
      if (depth === 1) {
        i += 1;
        continue;
      }
    } else if (c === ")") {
      depth -= 1;
      if (depth === 0) return out;
    }
    out += c;
    i += 1;
  }
  return out;
}

const ESCAPES = { n: "\n", r: "\r", t: "\t", b: "\b", f: "\f" };

export function unescapeString(s) {
  let out = "";
  let i = 0;
  while (i < s.length) {
    if (s[i] !== "\\") {
      out += s[i];
      i += 1;
      continue;
    }
    const next = s[i + 1];
    if (next >= "0" && next <= "9") {
      let digits = "";
      let j = i + 1;
      while (j < s.length && digits.length < 3 && s[j] >= "0" && s[j] <= "9") {
        digits += s[j];
        j += 1;
      }
      out += String.fromCharCode(parseInt(digits, 8) & 0xff);
      i = j;
    } else if (next in ESCAPES) {
      out += ESCAPES[next];
      i += 2;
    } else {
      out += next;
      i += 2;
    }
  }
  return out;
}

export async function load(bytes) {
  const pdf = new Pdf(bytes);
  await pdf.decodeAll();
  return pdf;
}

File: clo2plt/pdfread.py
"""Minimal read-only PDF access: object index, dictionary parsing, stream inflation.

Scoped deliberately to what CLO 3D emits (PDF 1.4, no object streams, FlateDecode
only). Anything outside that raises rather than guessing, so a future export that
breaks these assumptions fails loudly instead of producing a subtly wrong marker.
"""

import re
import zlib

_OBJ = re.compile(rb"(\d+)\s+(\d+)\s+obj\b")
_NUM = rb"[-+]?(?:\d+\.?\d*|\.\d+)(?:[eE][-+]?\d+)?"


class PdfError(Exception):
    pass


class Pdf:
    def __init__(self, data: bytes):
        if not data.startswith(b"%PDF-"):
            raise PdfError("not a PDF (missing %PDF- header)")
        if b"/ObjStm" in data:
            raise PdfError(
                "PDF uses object streams, which this reader does not support. "
                "Re-export from CLO 3D, or convert first with: gs -sDEVICE=pdfwrite "
                "-dCompatibilityLevel=1.4 -o flat.pdf input.pdf"
            )
        self.data = data
        self._offsets = {}
        for m in _OBJ.finditer(data):
            # Later definitions win: incremental updates append revised objects.
            self._offsets[int(m.group(1))] = m.end()

    def raw(self, num: int) -> bytes:
        """Bytes of object `num`, from after 'obj' up to 'endobj'."""
        try:
            start = self._offsets[num]
        except KeyError:
            raise PdfError(f"object {num} not found")
        end = self.data.find(b"endobj", start)
        if end < 0:
            raise PdfError(f"object {num} has no endobj")
        return self.data[start:end]

    def stream(self, num: int) -> bytes:
        """Decoded stream payload of object `num`."""
        body = self.raw(num)
        i = body.find(b"stream")
        if i < 0:
            raise PdfError(f"object {num} has no stream")
        j = i + len(b"stream")
        if body[j : j + 2] == b"\r\n":
            j += 2
        elif body[j : j + 1] in (b"\r", b"\n"):
            j += 1
        k = body.rfind(b"endstream")
        payload = body[j:k] if k > 0 else body[j:]
        header = body[:i]
        if b"/FlateDecode" not in header:
            if b"/Filter" in header:
                raise PdfError(f"object {num}: unsupported stream filter")
            return payload
        try:
            # decompressobj tolerates the trailing EOL some writers leave behind.
            return zlib.decompressobj().decompress(payload)
        except zlib.error as exc:
            raise PdfError(f"object {num}: inflate failed ({exc})")

    # -- dictionary helpers -------------------------------------------------
    # These read the raw object text rather than building a full object model.
    # CLO's dictionaries are flat and machine-generated, so targeted extraction
    # is both sufficient and far less code than a general parser.

    def ref(self, body: bytes, key: str):
        """Resolve `/Key N 0 R` to the integer object number, or None."""
        m = re.search(rb"/" + key.encode() + rb"\s+(\d+)\s+\d+\s+R\b", body)
        return int(m.group(1)) if m else None

    def name(self, body: bytes, key: str):
        """Read `/Key(literal string)`, unescaping PDF octal escapes."""
        m = re.search(rb"/" + key.encode() + rb"\s*\(", body)
        if not m:
            return None
        return unescape(_balanced_string(body, m.end() - 1)).decode("latin-1")

    def numbers(self, body: bytes, key: str):
        """Read `/Key[ n n n ]` as a list of floats."""
        m = re.search(rb"/" + key.encode() + rb"\s*\[([^\]]*)\]", body)
        if not m:
            return None
        return [float(x) for x in re.findall(_NUM, m.group(1))]

    def xobjects(self, body: bytes) -> dict:
        """Map `/Name N 0 R` entries of an XObject dictionary to object numbers."""
        return {
            k.decode("latin-1"): int(v)
            for k, v in re.findall(rb"/([A-Za-z0-9_.#-]+)\s+(\d+)\s+\d+\s+R", body)
        }


def _balanced_string(body: bytes, start: int) -> bytes:
    """Extract a PDF literal string starting at '(' honouring nesting/escapes."""
    depth = 0
    i = start
    out = bytearray()
    while i < len(body):
        c = body[i : i + 1]
        if c == b"\\":
            out += body[i : i + 2]
            i += 2
            continue
        if c == b"(":
            depth += 1
            if depth == 1:
                i += 1
                continue
        elif c == b")":
            depth -= 1
            if depth == 0:
                return bytes(out)
        out += c
        i += 1
    return bytes(out)


_ESC = {b"n": b"\n", b"r": b"\r", b"t": b"\t", b"b": b"\b", b"f": b"\f"}


def unescape(s: bytes) -> bytes:
    """Resolve PDF string escapes, including \\ddd octal."""
    out = bytearray()
    i = 0
    while i < len(s):
        if s[i : i + 1] != b"\\":
            out += s[i : i + 1]
            i += 1
            continue
        nxt = s[i + 1 : i + 2]
        if nxt.isdigit():
            j = i + 1
            digits = b""
            while j < len(s) and len(digits) < 3 and s[j : j + 1].isdigit():
                digits += s[j : j + 1]
                j += 1
            out.append(int(digits, 8) & 0xFF)
            i = j
        elif nxt in _ESC:
            out += _ESC[nxt]
            i += 2
        else:
            out += nxt
            i += 2
    return bytes(out)


def load(path: str) -> Pdf:
    with open(path, "rb") as fh:
        return Pdf(fh.read())
